streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamFlatTransform.java

```diff
@@ -52,11 +54,12 @@ public void init(final ProcessorContext context) {
         }
 
         @Override
-        public void process(final K1 key, final V1 value) {
-            final KeyValue<? extends K2, ? extends V2> pair = transformer.transform(key, value);
-
-            if (pair != null)
+        public void process(final KIn key, final VIn value) {
+            final Iterable<KeyValue<KOut, VOut>> pairs = transformer.transform(key, value);
+            Objects.requireNonNull(pairs, "result of transform can't be null");
```

Comment:
FWIW, I share your dissatisfaction with all the nullability in our API. It would be nice to migrate away from it over time.

Not important for this PR, but just to share the thought, as opposed to an optional return type, we could have coalesced transform and flattransform into one API by making the return type `void` and forcing implementers to use `context.forward`. Maybe this would have been better, since it would remove any ambiguity about how to use the API.

I think for that to work really well, we'd need to have a tighter control on the forwarded types (so we can still tell whether or not the key is going to change, and so we can properly type the down-stream). It's something I've been wanting to do anyway, to reduce casting. So it might be something to consider in the future.

Reply:
> making the return type `void` and forcing implementers to use `context.forward`

I don't think that is a good idea. Note that `context.forward` is not type safe by design. Adding return types allows us to offer a type safe way to return output records to Kafka Streams runtime.

Reply:
Yep, that's what I was referring to in the last paragraph. I think we could offer a different context API that offers the type safety we need, but it's speculative. I haven't actually tried. My comment wasn't for this PR at all, just thinking about where we could go in the future within the context of this change.

Note that what you point out about the type-danger of `context.forward` is applicable to transform today. We encourage users to use `forward` within their transformers, but doing so completely circumvents the type system.